Added netmap and state update requirements (#193)
leaningtech/labs: sites/cheerpx/src/content/docs/12-reference/00-CheerpX-Linux-create.md

This option configures network settings, which allows CheerpX to communicate over networks. For more detailed information about how CheerpX handles networking, including the use of Tailscale and overcoming browser limitations, see the [Networking](/docs/guides/Networking) guide.

### `authKey`

```ts
authKey?: string;
```

> The `authKey` is an optional string used for authentication and should be passed directly in the `create` call.

Reviewer: Reword this to be more similar to the brief explanation in the networking guide above, it's more accurate

Reviewer: Please remove the "passed directly into" part. It's both inaccurate (since it's passed actually inside an object) and redundant since this whole page documents the `create` API

Example:

```js
const cx = await CheerpX.Linux.create({
	mounts: mountPoints,
	networkInterface: { authKey: "YOUR KEY" },
});
```

> In this example, the `authKey` is included in the `networkInterface` object to enable authentication for the CheerpX instance.

Reviewer: ```suggestion
In this example, the `authKey` is included in the `networkInterface` object to automatically authenticate to Tailscale.
```

### `controlUrl`

```ts
controlUrl?: string;
```

> The `controlUrl` is an optional string used to specify the URL of a self-hosted server.

Reviewer: ```suggestion
The `controlUrl` is an optional string used to specify the URL of a self-hosted Headscale server.
```

Also link back to the relevant guide. The short description in the Networking guide it's also better here

Example:

```js
const cx = await CheerpX.Linux.create({
	mounts: mountPoints,
	networkInterface: { controlUrl: "YOUR URL" },
});
```

> Here, the `controlUrl` is used to point the CheerpX instance to a custom self-hosted server.

Reviewer: Consider removing these extra comment lines, they are mostly redundant. The example is enough.

Reviewer: They can be left if there is actually something to add after the example code

### `loginUrlCb`

```ts
loginUrlCb?: (url: string) => void;
```

> The `loginUrlCb` is a callback function that handles login URLs during the authentication process. It provides the URL needed to continue the login process, such as authenticating with Tailscale.

Reviewer: ```suggestion
The `loginUrlCb` is a callback function that handles login URLs during the authentication process. It receives the URL that should be visited to continue the login process. This is necessary when authenticating with Tailscale.
```

Example:

```js
const cx = await CheerpX.Linux.create({
	mounts: mountPoints,
	networkInterface: { loginUrlCb },
});

function loginUrlCb(url) {
	console.log("Login URL is ready:", url);
	// Open the login URL in a new tab or window
	window.open(url, "_blank");
}
```

> The `loginUrlCb` provides the user with a login URL for authentication. It logs the URL and opens it in a new tab or window. You can also adapt it to display the URL within your app or handle it differently as needed.

Reviewer: ```suggestion
The `loginUrlCb` receives a login URL for authentication. It logs the URL and opens it in a new tab or window. You can also adapt it to display the URL within your app or handle it differently as needed.
```

### stateUpdateCb

```ts
stateUpdateCb?: (state: number) => void;
```

The `stateUpdateCb` is a callback that runs when the connection state changes. The `state` parameter represents the connection status.

Example:

```js
const cx = await CheerpX.Linux.create({
	mounts: mountPoints,
	networkInterface: { stateUpdateCb },
});

function stateUpdateCb(state) {
	if (state === 6) {
		console.log("Connected");
	}
}
```

This example checks if the `state` indicates a "connected" status (`state = 6`) and logs "Connected" when the connection is established.

### netmapUpdateCb

```ts
netmapUpdateCb?: (map: any) => void;
```

> The `netmapUpdateCb` is a callback triggered whenever the network configuration updates. It provides details about the current network configuration.

Reviewer: ```suggestion
The `netmapUpdateCb` is a callback that runs whenever the network configuration updates. It provides details about the current network configuration.
```

Be consistent with the previous one

[Promise]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

### `authKey`

```ts
authKey?: string;
```

The `authKey` is an optional string used for authentication. It is usually taken from the URL, allowing the CheerpX Linux environment to be set up automatically without the user needing to enter it manually.

**Example of usage:**

```js
let authKey = undefined;
if (browser) {
	let params = new URLSearchParams("?" + window.location.hash.substr(1));
	authKey = params.get("authKey") || undefined;
}
const networkInterface = { authKey };
```

**Explanation**

1. The `authKey` is taken from the URL, allowing the application to automatically use the authentication key provided by the browser.

2. The `authKey` is added to the `networkInterface` object, ensuring the application can access the authentication data whenever needed in the CheerpX environment.

### `controlUrl`

```ts
controlUrl?: string;
```

> The `controlUrl` is an optional string parameter representing a URL used to manage or monitor the connection. It is typically fetched from the URL parameters and can be utilized for dashboard or configuration purposes.

Reviewer: Same as before. Simplify this in general. The purpose of this option is to select the self-hosted server, all the rest is noise.

In general prefer explaining what is the purpose of options from the user point of view,. Nobody cares about "the control plain", people just want to self host somehow

**Example of usage:**

```js
let controlUrl = undefined;
if (browser) {
	let params = new URLSearchParams("?" + window.location.hash.substr(1));
	controlUrl = params.get("controlUrl") || undefined;
}
const dashboardUrl = controlUrl
	? null
	: "https://login.tailscale.com/admin/machines";
const networkInterface = { controlUrl };
```

**Explanation**

1. The `controlUrl` is taken from the URL and added to the `networkInterface` object, making it easy to connect to the appropriate control dashboard.

2. If `controlUrl` isn’t provided, the app automatically uses a default URL (`dashboardUrl`) for monitoring or setup.

### `loginUrlCb`

```ts
loginUrlCb?: (url: string) => void;
```

The `loginUrlCb` is a callback function designed to handle login URLs during the authentication process. It plays a critical role in the connection flow by setting the application's state to indicate readiness for login and resolving the URL required for the login process.

**Example of usage:**

```js
function loginUrlCb(url) {

> 	connectionState.set("LOGINREADY");

Reviewer: The use of WebVM/Svelte specific code is not a good idea, since people will most likely not have this setup. It's not strictly required so it can be removed.

> 	resolveLogin(url);

Reviewer: This is another WebVM idiom, we do this to update the networking login button when the login URL is ready, but this is a detail. Here we should strive for minimal self-contained examples. Printing the URL to the console or using window.open are better solution. I strongly recommend writing down a complete example to make sure everything works as intended.

```js
resolveLogin(url);
```

**How it works**

The `loginUrlCb` is part of the `networkInterface` object, which organizes the key callbacks and functions for managing the network state. It interacts with:

- `startLogin`: The function that initiates the login process. It depends on the `loginUrlCb` to resolve the login URL and update networkData.

```js
export async function startLogin() {
	connectionState.set("LOGINSTARTING");
	const url = await loginPromise; // Waits for `loginUrlCb` to resolve the URL.
	networkData.loginUrl = url; // Stores the resolved URL for application-wide access.
	return url;
}
```

### StateUpdateCb

```ts
stateUpdateCb?: (state: number) => void;
```

This callback function is invoked whenever the connection state changes, enabling you to react dynamically to transitions such as connecting, disconnecting, or successfully establishing a connection. The `state` parameter is an integer representing the current state.

**Example of usage:**

```js
function stateUpdateCb(state) {
	switch (state) {
		case 6: // Running

> 			connectionState.set("CONNECTED");

Reviewer: Remove WebVM specific code

			break;
		default:
			connectionState.set("DISCONNECTED");
	}
}
```

**Explanation**

- **State** `6` (Running): The callback sets the application's `connectionState` to `"CONNECTED"`, indicating the system is online.
- **Other states**: The default case handles all non-running states, setting `connectionState` to `"DISCONNECTED"`.

**How it works**

The `stateUpdateCb` is part of the `networkInterface` object and is triggered internally during connection events. It enables you to synchronize the application's user interface or internal state with the current connection status.

### NetmapUpdateCb

```ts
netmapUpdateCb?: (map: any) => void;
```

This callback is triggered whenever the network map is updated. The `map` parameter contains the full network configuration, including information about the current IP address and peers.

**Example of usage:**

```ts
function netmapUpdateCb(map) {
	networkData.currentIp = map.self.addresses[0];
	var exitNodeFound = false;
	for (var i = 0; i < map.peers.length; i++) {
		if (map.peers[i].exitNode) {
			exitNodeFound = true;
			break;
		}
	}
	if (exitNodeFound) {

> 		exitNode.set(true);

Reviewer: Remove WebVM specific code
